src/unchunker.ts

```diff
@@ -1,247 +1,410 @@
+    private queueUnorderedChunk(chunk: Chunk): boolean {
+        // Append chunk
+        this.queuedChunksTotalByteLength += chunk.payload.byteLength;
+        if (this.queuedChunks === null) {
+            this.queuedChunks = [chunk];
+            return false;
         }
+        this.queuedChunks.push(chunk);
 
-        // Sort chunks
-        this.chunks.sort((a: Chunk, b: Chunk) => {
+        // Sort chunk queue
+        this.queuedChunks.sort((a: Chunk, b: Chunk) => {
             if (a.serial < b.serial) {
                 return -1;
-            } else if (a.serial > b.serial) {
+            }
+            if (a.serial > b.serial) {
                 return 1;
             }
             return 0;
         });
 
-        // Allocate buffer
-        const capacity = this.chunks[0].data.byteLength * this.messageLength;
-        const buf = new Uint8Array(new ArrayBuffer(capacity));
-
-        // Add chunks to buffer
-        let offset = 0;
-        const firstSize = this.chunks[0].data.byteLength;
-        const contextList = [];
-        for (const chunk of this.chunks) {
-            if (chunk.data.byteLength > firstSize) {
-                throw new Error('No chunk may be larger than the first chunk of that message.');
-            }
-            buf.set(chunk.data, offset);
-            offset += chunk.data.length;
-            if (chunk.context !== undefined) {
-                contextList.push(chunk.context);
+        // Check if ready
+        const iterator = this.queuedChunks.values();
+        let previousChunk = iterator.next().value;
+        if (previousChunk.serial !== this.nextOrderedSerial) {
+            return false;
+        }
+        for (const currentChunk of iterator) {
+            if (previousChunk.serial + 1 !== currentChunk.serial) {
+                return false;
             }
+            previousChunk = currentChunk;
         }
+        return true;
+    }
 
-        // Return result object
-        return {
-            message: buf.slice(0, offset),
-            context: contextList,
-        };
+    /**
+     * Moves the queued chunks to the contiguous buffer.
+     *
+     * Should be called once the queue contains consecutive chunks and there is
+     * no gap between the contiguous chunk buffer and our queued chunks.
+     */
+    private moveQueuedChunks(): void {
+        const chunk = this.contiguousChunks.addBatched(this.queuedChunks, this.queuedChunksTotalByteLength);
+        // Note: `chunk` is the last chunk in the sequence and has the highest serial number
+        this.nextOrderedSerial = chunk.serial + 1;
+        this.queuedChunks = null;
     }
 
     /**
-     * Return whether last chunk is older than the specified number of miliseconds.
+     * Get the reassembled message.
+     *
+     * @return The completed message.
+     */
+    public getMessage(): Uint8Array {
+        return this.contiguousChunks.getMessage();
+    }
```

Comment:
All usages of it check for completeness before calling it.

Reply:
Alright, I'll add a *complete* check since it's cheap. But if you want to knock yourself out with defensive programming, you would also have to check the chunk id every time and I'm not going that way (and neither did you).